provider/tests/provider_test.go

```diff
@@ -0,0 +1,276 @@
+// bucket to cleanup.
+func TestTags(t *testing.T) {
+	replaySequence(t, `[
+  {
+    "method": "/pulumirpc.ResourceProvider/Configure",
+    "request": {
+      "variables": {
+        "aws:config:defaultTags": "{\"tags\":{\"abc\":\"def\",\"foo\":\"buzz\"}}",
+        "aws:config:region": "us-west-2",
+        "aws:config:skipCredentialsValidation": "false",
+        "aws:config:skipMetadataApiCheck": "true",
+        "aws:config:skipRegionValidation": "true"
+      },
+      "args": {
+        "defaultTags": "{\"tags\":{\"abc\":\"def\",\"foo\":\"buzz\"}}",
+        "region": "us-west-2",
+        "skipCredentialsValidation": "true",
+        "skipMetadataApiCheck": "true",
+        "skipRegionValidation": "true"
+      },
+      "acceptSecrets": true,
+      "acceptResources": true,
+      "sendsOldInputs": true
+    },
+    "response": {
+      "supportsPreview": true
+    },
+    "metadata": {
+      "kind": "resource",
+      "mode": "client",
+      "name": "aws"
+    }
+  },
+  {
+    "method": "/pulumirpc.ResourceProvider/Check",
+    "request": {
+      "urn": "urn:pulumi:test::secret-random-yaml::aws:s3/bucketV2:BucketV2::b",
+      "olds": {},
+      "news": {
+        "tags": {
+          "foo": "fizz"
+        }
+      },
+      "randomSeed": "6ba2bL+cZxmiU0dbBQuYcRtIJBjlR6W2hOyKGiafB3Y="
+    },
+    "response": {
+      "inputs": {
+        "__defaults": [
+          "bucket",
+          "forceDestroy"
+        ],
+        "bucket": "b-5b730b2",
+        "forceDestroy": false,
+        "tags": {
+          "__defaults": [],
+          "abc": "def",
+          "foo": "fizz"
+        }
+      }
+    },
+    "metadata": {
+      "kind": "resource",
+      "mode": "client",
+      "name": "aws"
+    }
+  },
+  {
+    "method": "/pulumirpc.ResourceProvider/Create",
+    "request": {
+      "urn": "urn:pulumi:test::secret-random-yaml::aws:s3/bucketV2:BucketV2::b",
+      "properties": {
+        "__defaults": [
+          "bucket",
+          "forceDestroy"
+        ],
+        "bucket": "b-5b730b2",
+        "forceDestroy": false,
+        "tags": {
+          "__defaults": [],
+          "abc": "def",
+          "foo": "fizz"
+        }
+      },
+      "preview": true
+    },
+    "response": {
+      "properties": {
+        "bucket": "b-5b730b2",
+        "forceDestroy": false,
+        "id": "",
+        "tags": {
+          "abc": "def",
+          "foo": "fizz"
+        },
+        "tagsAll": {
+          "abc": "def",
+          "foo": "fizz"
+        }
+      }
+    },
+    "metadata": {
+      "kind": "resource",
+      "mode": "client",
+      "name": "aws"
+    }
+  }
+]`)
+}
+
+// Repro https://github.com/pulumi/pulumi-aws/issues/2633
+func TestRegression2633(t *testing.T) {
+	t.Parallel()
+	t.Run("changed tags", func(t *testing.T) { replaySequence(t, regress2633(true)) })
+	t.Run("removed tags", func(t *testing.T) { replaySequence(t, regress2633(false)) })
+}
+
+func regress2633(hasTags bool) string {
```

Comment:
Can you walk me through this? I don't understand. Suppose the scenario is hasTags=true.

The provider is configured with defaultTags having "foo" set.

Why is the diff deciding that "detailedDiff": { "tags.foo": { "kind": "DELETE" } }?

Shouldn't tags.foo be propagated from provider defaults and be stable?

Reply:
That test is outdated. It is not run by CI. I have removed the test in https://github.com/pulumi/pulumi-aws/pull/2655/commits/cc12424c508051839607eee283419ef11860c7fe.